src/string.cr

```diff
@@ -1056,59 +1056,70 @@ class String
     end
   end
 
-  def unsafe_byte_at(index)
+  # Returns the byte at the given *index* without bounds checking.
+  def unsafe_byte_at(index : Int) : UInt8
     to_unsafe[index]
   end
 
-  # Returns a new `String` with each uppercase letter replaced with its lowercase
-  # counterpart.
+  # Returns a new `String` with each uppercase letter replaced with its lowercase counterpart.
   #
   # ```
   # "hEllO".downcase # => "hello"
   # ```
-  def downcase(options = Unicode::CaseOptions::None)
+  def downcase(options : Unicode::CaseOptions = :none) : String
     return self if empty?
 
+    String.build(bytesize) { |io| downcase io, options }
+  end
+
+  # Writes a downcased version of `self` to the given *io*.
+  #
+  # ```
+  # io = IO::Memory.new
+  # "hEllO".downcase io
+  # io.to_s # => hello
+  # ```
+  def downcase(io : IO, options : Unicode::CaseOptions = :none) : Nil
     if ascii_only? && (options.none? || options.ascii?)
-      String.new(bytesize) do |buffer|
-        bytesize.times do |i|
-          buffer[i] = to_unsafe[i].unsafe_chr.downcase.ord.to_u8
-        end
-        {@bytesize, @length}
+      bytesize.times do |i|
+        io.write_byte unsafe_byte_at(i).unsafe_chr.downcase.ord.to_u8
       end
     else
-      String.build(bytesize) do |io|
-        each_char do |char|
-          char.downcase(options) do |res|
-            io << res
-          end
+      each_char do |char|
+        char.downcase(options) do |res|
+          io << res
         end
       end
     end
   end
 
-  # Returns a new `String` with each lowercase letter replaced with its uppercase
-  # counterpart.
+  # Returns a new `String` with each lowercase letter replaced with its uppercase counterpart.
   #
   # ```
   # "hEllO".upcase # => "HELLO"
   # ```
-  def upcase(options = Unicode::CaseOptions::None)
+  def upcase(options : Unicode::CaseOptions = :none) : String
     return self if empty?
 
+    String.build(bytesize) { |io| upcase io, options }
```

Comment:
Same comment as before. I think the definitions need to be duplicated or a different reuse must be done to avoid losing that optimization.

Reply:
Is there a benefit in having the optimization in the IO overload as well?  I.e. if the main benefit is _how_ the string is created could we leave the optimization only in the non IO overload.  Then the IO overload could just use `eac_char`?

Reply:
I think only a benchmark can tell that. But given that we didn't have the IO overload before, using each_char is better than nothing.

Reply:
Looks like `#each_char` as an optimization for `ascii_only` strings, so probably the best we're going to get since it has to be written to an IO.